Objective-C/Tests/DatabaseTest.m

```diff
@@ -582,6 +582,57 @@ - (void) testSaveDocWithDeletedConflictUsingConcurrencyControl: (int)concurrency
     [self cleanDB];
 }
 
+- (void) testCancelConflictHandler {
+    NSString* docID = @"doc1";
+    CBLMutableDocument* doc = [[CBLMutableDocument alloc] initWithID: docID];
+    [doc setString: @"Tiger" forKey: @"firstName"];
+    [self saveDocument: doc];
+    AssertEqual([self.db documentWithID: docID].generation, 1u);
+    
+    CBLMutableDocument* doc1a = [[self.db documentWithID: docID] toMutable];
+    CBLMutableDocument* doc1b = [[self.db documentWithID: docID] toMutable];
+    
+    [doc1a setString: @"Scotty" forKey: @"nickName"];
+    [self saveDocument: doc1a];
+    AssertEqual([self.db documentWithID: docID].generation, 2u);
+    
+    NSError* error;
+    [doc1b setString: @"Scott" forKey: @"nickName"];
+    [self.db saveDocument: doc1b
+          conflictHandler:^BOOL(CBLMutableDocument * cur, CBLDocument * old) {
+              AssertEqualObjects(doc1b.toDictionary, cur.toDictionary);
+              AssertEqualObjects(doc1a.toDictionary, old.toDictionary);
+              return NO;
+          } error: &error];
+    AssertNil(error);
```

Comment:
Isn't `saveDocument` supposed to return a Conflict error if the handler returns NO?

Also, testing `error` is the wrong way to check for success/failure. In Cocoa APIs, the return value indicates success/failure. The NSError has a value only on failure; on success _its value is undefined_. (Yeah, you initialized it to nil, but it's possible that the method you called used it itself for other calls that failed, but it recovered from those and returned YES, but left the error non-nil.)

Reply:
will assert the `db saveDocument` method. But should we also make sure, the provided error is empty when success

Reply:
yea, this was wrong, checking error as empty. which is corrected.